### ISSUE TITLE
Profiling dynamic analysis

### ISSUE THREAD

**Issue by aherrmann** [facebook/buck2#707]
Buck2 supports profiling of Starlark as documented under [Observability and Optimization](https://buck2.build/docs/rule_authors/optimization/). In particular, it supports profiling of analysis via `buck2 profile analysis`. However, this feature only captures the static analysis. Any Starlark code that is executed as part of dynamic output actions is not included in this profile.

I would like to profile Starlark code that is executed as part of dynamic output actions, i.e. the [implementation of `f` in `ctx.actions.dynamic_output`](https://buck2.build/docs/api/build/actions/#actionsdynamic_output), or the equivalent under the [upcoming new API](https://buck2.build/docs/api/build/actions/#actionsdynamic_output_new). The motivation is that Buck2 reports a noticeable amount of time spent in "dynamic analysis" on the TUI for some of our targets, and I would like to better understand where this time is spent and optimize the relevant code. (Anecdotally, we had an instance where dynamic analysis took 20s. Luckily we were able to reduce it to 2s simply by refactoring, but we would like to continue optimizing it based on actual data.)

Alternative profiling methods, in particular [native profiling](https://buck2.build/docs/rule_authors/optimization/#native-profiling) did not yield meaningful information about the Starlark code that we need to optimize.

**Comment by stepancheg** [comment]
@aherrmann right. Internally we agreed to go with this approach — add `--profile-mode=` to many commands and output profile of whatever is evaluated. So far I have implemented this flag for cquery command (which is useful to do recursive "loading" profiler), but it is not committed yet for cquery command https://www.internalfb.com/diff/D59088660.

**Comment by aherrmann** [comment]
@stepancheg Thanks for the update that's good to hear. Thanks for implementing the feature for cquery. Unfortunately, I don't have access to the internal diff. Is it also visible on GitHub somewhere?

**Comment by alexlian** [comment]
@aherrmann 

> Unfortunately, I don't have access to the internal diff. Is it also visible on GitHub somewhere?

Indeed, it's this commit: https://github.com/facebook/buck2/commit/254d171d1699417aa0d8efbe02c9bd5bffdcf275

**Comment by alexlian** [comment]
@aherrmann 
Let us know if this isn't resolved. Closing for now, feel free to re-open or start a more specific issue.

**Comment by aherrmann** [comment]
@alexlian Thanks for the update and linking to the commit. I tried out `--profile-mode`. To my understanding this is currently only available on `cquery`, but not yet on the `build` command. However, IIUC the feature would need to be supported by the `build` command, or perhaps the `aquery` command, to provide profiling of the dynamic analysis.

I don't have permissions to re-open the issue myself. But, IIUC it should be.

**Comment by cjhopman** [comment]
I have changes I've been slowly landing that will resolve this. I didn't realize (or had forgotten) their were public requests for improving this, but I'll see about getting those in more quickly. 

Once they land, all starlark evaluations will have a string identifier and all commands will allow providing a regex to match against those and all matched evaluations will be profiled (they almost all already have the identifier, it's the one used for the thread name in the debugger).

**Comment by avdv** [comment]
Hey @cjhopman 

any update on this?

We are currently facing an issue with dynamic actions being a bottleneck, degrading over the last few months / weeks.

**Comment by lf-** [comment]
Copied from an internal chat for the benefit of others: try `buck2 build <target> --profile-patterns 'dynamic_output.*' --profile-patterns-mode time-flame --profile-patterns-output /tmp/prof`

**Comment by cjhopman** [comment]
Yeah, so all commands now have the `--profile-patterns` flags and the `--help` text for those explains this in more detail. But basically profile-patterns accepts a regex and all starlark executions have a string identifier that the regex will be matched against to enable or not profiling. And then we produce a bunch of profiles in the passed in output directory. using this, it should be possible to profile any starlark evaluation.

**Comment by lf-** [comment]
another fun one:
```
buck2 profile analysis '//:mytarget' --recursive --output=xxx-starlark-profile --mode=time-flame
```

i think this is resolved at this point, so closing.
